### DISCUSSION TITLE
Getting started w/Uno Platform

### DISCUSSION THREAD

**Discussion by casterle** [unoplatform/uno#7867] [Q&A] [answered]
TL;DR:

I'm trying Uno for the first time and need some help.

First, Uno-Check (Ver 1.0.2.0) fails as described below and may have left my Uno environment hosed as repairs seem to have failed as well.

Also, Uno-Check seems to have installed a bunch of preview sdk's that I didn't expect, including a .NET 6.0.200 preview SDK. When I'm learning something new, I prefer to avoid preview software so I can be somewhat sure that errors are mine rather than system bugs. Can I avoid preview software while writing Uno programs? I only want to target desktops...

I recently upgraded to Win11 Pro x64 (a clean install) and have finally gotten my system and programs configured and ready to start a Uno project.

Running Uno-Check leads to a series of errors as follows. After passing several initial tests, Uno-Check fails due to missing SDKs:

```
android-aot (Microsoft.NET.Sdk.Android.Manifest-6.0.100 : 31.0.101-preview.11.81) ios (Microsoft.NET.Sdk.iOS.Manifest-6.0.100 : 15.0.101-preview.11.391)

maccatalyst (Microsoft.NET.Sdk.MacCatalyst.Manifest-6.0.100 : 15.0.101-preview.11.391)

tvos (Microsoft.NET.Sdk.tvOS.Manifest-6.0.100 : 15.0.101-preview.11.391)

macos (Microsoft.NET.Sdk.macOS.Manifest-6.0.100 : 12.0.101-preview.11.391)

maui (Microsoft.NET.Sdk.Maui.Manifest-6.0.100 : 6.0.101-preview.10.2068)

```
When I tell the program to attempt to fix the issues, I get:

```
No workloads installed for this feature band. To update workloads installed with earlier SDK versions, include the --from-previous-sdk option.

Failed to update the advertising manifest microsoft.net.workload.emscripten: microsoft.net.workload.emscripten.manifest-6.0.200 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json, https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json, https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json, https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json, https://api.nuget.org/v3/index.json"..

Failed to update the advertising manifest microsoft.net.workload.mono.toolchain: Failed to validate package signing.

Verifying Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.200.6.0.2-mauipre.1.22068.7error: NU3004: The package is not signed.Package signature validation failed.

Updated advertising manifest microsoft.net.sdk.tvos.Downloading microsoft.net.sdk.android.manifest-6.0.200.msi.x64 (31.0.101-preview.11.81)Workload installation failed. Rolling back installed packs...Downloading microsoft.net.sdk.android.manifest-6.0.200.msi.x64 (31.0.101-preview.11.117)Installation rollback failed: One or more errors occurred. (microsoft.net.sdk.android.manifest-6.0.200.msi.x64::31.0.101-preview.11.117 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json;https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json;https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json;https://api.nuget.org/v3/index.json".)

Workload update failed: One or more errors occurred.(microsoft.net.sdk.android.manifest-6.0.200.msi.x64::31.0.101-preview.11.81 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json;https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json;https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json;https://api.nuget.org/v3/index.json".)

Fix failed - Workload Update failed: `dotnet workload update--no-cache--disable-parallel--from-rollback-file "C:\Users\casterle\AppData\Local\Temp\maui-check-2a115ce0\workload.json"--source "https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json"--source "https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json"--source "https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json"--source "https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json"--source "https://api.nuget.org/v3/index.json"`! There were one or more problems detected.Please review the errors and correct them and run uno-check again.
```

I'm not sure what most of these errors mean, much less how to fix them. Given the first error indicated that I was missing sdk's and suggested a fix, I inserted "--from-previous-sdk" into the options into the Uno-Check-generated command line and tried again:

The following was entered (on a single line):

```
dotnet workload update 
--no-cache 
--disable-parallel 
--from-previous-sdk 
--from-rollback-file "C:\Users\casterle\AppData\Local\Temp\maui-check-2a115ce0\workload.json" 
--source "https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json" 
--source "https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json" 
--source "https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json" 
--source "https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json" 
--source "https://api.nuget.org/v3/index.json"
```

Which resulted in:

```
Failed to update the advertising manifest microsoft.net.workload.emscripten: microsoft.net.workload.emscripten.manifest-6.0.200 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json, https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json, https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json, https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json, https://api.nuget.org/v3/index.json"..

Failed to update the advertising manifest microsoft.net.workload.mono.toolchain:Failed to validate package signing.

Verifying Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.200.6.0.2-mauipre.1.22068.7error: NU3004: The package is not signed.Package signature validation failed.

Downloading microsoft.net.sdk.android.manifest-6.0.200.msi.x64 (31.0.101-preview.11.81)Workload installation failed. Rolling back installed packs...Downloading microsoft.net.sdk.android.manifest-6.0.200.msi.x64 (31.0.101-preview.11.117)Installation rollback failed: One or more errors occurred.(microsoft.net.sdk.android.manifest-6.0.200.msi.x64::31.0.101-preview.11.117 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json;https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json;https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json;https://api.nuget.org/v3/index.json".)

Workload update failed: One or more errors occurred. (microsoft.net.sdk.android.manifest-6.0.200.msi.x64::31.0.101-preview.11.81 is not found in NuGet feeds https://pkgs.dev.azure.com/xamarin/public/_packaging/maui-main/nuget/v3/index.json;https://pkgs.dev.azure.com/azure-public/vside/_packaging/xamarin-impl/nuget/v3/index.json;https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json;https://dnceng.pkgs.visualstudio.com/public/_packaging/dotnet6/nuget/v3/index.json;https://api.nuget.org/v3/index.json".)
```

Running Uno-Check again fails to find the sdk's it initially complained about, so adding the '--from-previous-sdk' flag didn't fix the problem.

FYI, dotnet --info reveals:

```
PS C:\Windows\system32> dotnet --info
.NET SDK (reflecting any global.json):
 Version:   6.0.200-preview.21617.4
 Commit:    58f9685dd5

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22000
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.200-preview.21617.4\

Host (useful for support):
  Version: 6.0.1
  Commit:  3a25a7f1cc

.NET SDKs installed:
  5.0.404 [C:\Program Files\dotnet\sdk]
  6.0.101 [C:\Program Files\dotnet\sdk]
  6.0.200-preview.21617.4 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.22 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.13 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.1 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
```
So it appears that Uno-Check uninstalled some components and was unable to restore them when the upgrade failed.

Can someone please help me out?

**Comment by limeniye** [comment, marked as answer]
Good day! 
Can you try this?

https://github.com/unoplatform/uno/discussions/7833

**Reply by casterle** [reply]
Thank you, that made the necessary repairs. Subsequently running Uno-Check without the recommended --manifest argument still fails, but running again it with the --manifest argument succeeded without making repairs this time.

How do I know what --manifest argument to use in the future? Is this documented somewhere?

As someone just trying to get started with Uno (and X-Platform in general), Uno seems a bit of a mess...

**Reply by jeromelaban** [reply]
.NET 6 and Visual Studio Previews are changing very frequently lately, this means that uno is in preview for those areas as a result. If you want to have a stable experience, use stable bits for both with Visual Studio 17.0.x.

**Reply by casterle** [reply]
Are there different versions of Uno installed for the release and preview versions of VS? Uno-Check didn't ask which version of VS I wanted to install Uno for and just installed the pre-release stuff...

I intended to use Uno with the release version, and Maui (which I'd also like to investigate) with the preview version.

**Reply by drosenba** [reply]
I'm new to Uno, and quite confused as how to review a Mobile (Android) app...
In VS 2022, mobile is "incompatible"
[https://stackoverflow.com/questions/70561181/app-mobile-project-appears-as-incompatible](url)
But when I try to run VS 2022 Preview (3 and 4), I run into this issue
[https://github.com/unoplatform/uno/discussions/7648](url)

even though I tried to apply uno updates
[https://github.com/unoplatform/uno/discussions/7833](url)

which initially fails
   Verifying Microsoft.NET.Workload.Mono.ToolChain.Manifest-6.0.200.6.0.2-mauipre.1.22076.3
   error: NU3004: The package is not signed.

Although re-running the same   uno-check --manifest [...]   then tells me all is OK.

I don't mind falling back to a prior version (of either Uno or VS) if I have to, as for now I just want to see want the controls look like in the Android simulator.   Please advise.

**Reply by limeniye** [reply]
@drosenba  did u install Android SDK? 
Tools -> Android -> AndroidSDK

**Reply by drosenba** [reply]
Yes, I have Android SDK installed.  I noticed that the Android 11.0 - R was missing some items that were in the 12.0-S, so I installed them.  Then I created a new Uno project from template.  Many errors come up, but I've seen that before, so I close VS, and then reopen it, and then do NuGet update of all packages needing update (except for ones related to Microsoft Logging, as per a getting started blog).  These are the errors that appear now
![Capture_ErrorList](https://user-images.githubusercontent.com/13946985/151461366-2e5322dc-c44a-405a-bd9c-d534bc262b9e.PNG)
.

**Reply by limeniye** [reply]
@drosenba I had the same problem.
- I installed last SDK
https://developer.microsoft.com/ru-ru/windows/downloads/sdk-archive/
- and I have used uno-check from this post
https://github.com/unoplatform/uno/discussions/7833

**Reply by limeniye** [reply]
also you can just delete this platforms from project-properties.

**Reply by drosenba** [reply]
After installing windows SDK as per above, I re-run the 
> uno-check --manifest https://raw.githubusercontent.com/unoplatform/uno.check/209d531333ecc43eeec048e60fb62f679624b0e6/manifests/uno.ui-preview.manifest.json

And as per before, I again get this error
> Workload installation failed: One or more errors occurred. (Failed to validate package signing.
> Verifying Microsoft.NETCore.App.Runtime.AOT.win-x64.Cross.android-x86.Msi.x64.6.0.2-mauipre.1.22054.8
> error: NU3004: The package is not signed.
> Package signature validation failed.

How can this be resolved?

**Reply by jeromelaban** [reply]
Guidelines are updated for Preview 3 and 4: https://github.com/unoplatform/uno/discussions/7833

**Reply by drosenba** [reply]
I repeated the "dotnet tool update..." and ".uno-check --manifest ..", and I still get the above error "Package signature validation failed."

**Reply by jeromelaban** [reply]
The manifest and version have changed, make sure to update to the latest that is mentioned in the other discussion.

**Reply by drosenba** [reply]
As per the other discussion, ...

I did this
> dotnet tool update -g uno.check --version 1.1.0-dev.22

and then this
>uno-check --manifest https://raw.githubusercontent.com/unoplatform/uno.check/4b89f56d8fa3f2cdd980425dcc8cbc4db1249d99/manifests/uno.ui-preview.manifest.json

Now I get this
> ! There were one or more problems detected.
> Please review the errors and correct them and run uno-check again.

But no errors are reported.

**Reply by jeromelaban** [reply]
You're likely getting this error: https://github.com/unoplatform/uno.check/issues/11

**Reply by drosenba** [reply]
I ran the uno-check with --verbose

> Android Emulator Checkup...
System.NullReferenceException: Object reference not set to an instance of an object.
  at DotNetCheck.AndroidSdk.AvdManager.ListAvds() in /_/UnoCheck/AndroidSdk/AvdManager.cs:183
  at DotNetCheck.Checkups.AndroidEmulatorCheckup.Examine(SharedState history) in /_/UnoCheck/Checkups/
     AndroidEmulatorCheckup.cs:51
  at DotNetCheck.Cli.CheckCommand.ExecuteAsync(CommandContext context, CheckSettings settings) in /_/UnoCheck/
     CheckCommand.cs:167


(My Android Emulator is running.  On its "screen" header, it says "... - pixel_5_-api_30:5554" )

**Reply by jeromelaban** [reply]
This last issue is going to be fixed by https://github.com/unoplatform/uno.check/pull/27. If you don't have any other errors, you should be good to go.

**Reply by drosenba** [reply]
I don't know what you mean by "should be good to go".  Now, or after the fix?

**Reply by jeromelaban** [reply]
If you don't have any other errors, everything else should be working.

**Reply by drosenba** [reply]
I opened VS 2022  17.1.0 Preview 4.0.  I create a new project "Multi-Platform App (Uno Platform|net6).  I still get lots of errors and warnings, and cannot run the .Mobile project.  

In the Error List, I get errors like this
> Severity	Code	Description	Project	File	Line	Suppression State
> 
> Error	NU1012	Platform version is not present for one or more target frameworks, even though they have specified a platform: net6.0-macos	UnoPreviewTest1.Mobile	C:\Users\dennis\source\repos\UnoPreviewTest1\UnoPreviewTest1\UnoPreviewTest1.Mobile\UnoPreviewTest1.Mobile.csproj	1	
> 
> Error	XA1018	Specified AndroidManifest file does not exist: C:\Users\dennis\source\repos\UnoPreviewTest1\UnoPreviewTest1\UnoPreviewTest1.Mobile\AndroidManifest.xml.	UnoPreviewTest1.Mobile	C:\Program Files\dotnet\packs\Microsoft.Android.Sdk.Windows\31.0.200-preview.12.2\tools\Xamarin.Android.Common.targets	515	

When I build, the Output tab's "Show output from" **Build**, shows many errors like this
3>CSC : error CS2015: 'C:\Users\dennis\.nuget\packages\uno.ui\4.0.11\analyzers\dotnet\cs\Mono.Cecil.Mdb.dll' **is a binary file instead of a text file**

Also, in the Output tab's from "Uno Platform", I see this
[ERROR] The operation failed as details for project UnoPreviewTest1.Skia.Gtk could not be loaded.
